### ISSUE TITLE
How about a monorepo 😄 

### ISSUE THREAD

**Issue by derberg** [asyncapi/generator#1044]
## Current state

- https://github.com/asyncapi/generator/ that is a core
- https://github.com/asyncapi/generator-react-sdk for react render engine
- https://github.com/asyncapi/generator-filters/ with shared filters
- https://github.com/asyncapi/generator-hooks/ with shared hooks
- Multiple repositories with templates

This granularity is technically possible so templated development is flexible. 
**But the fact that we make it possible for others, to easily create private, independent and use case specific templates doesn't mean we should have this level of granularity in asyncapi GH org**

## Proposal - discussion

How about we turn `generator` into a monorepo?

- `generator-react-sdk` - not a separate npm package, just natively part of generator
- `generator-filters` and `generator-hooks` also natively a part of generator (we still want to enable people to provide theirs independently)
- yes, templates as part of generator repo, but in case of templates, I guess we would need to publish separate packages 🤔 although I'm not 100% sure we need to

## Why

- having to use `generator-react-sdk` as separate package is kinda confusing and in general my impression is that motivation for having it as a separate package was that then it can be written in TS
- over last 2 year we had zero improvements in hooks and filters, they are hidden small repos and template developers basically work on their own filters and hooks in their template repos. 
- it would not be obligatory to move `@asyncapi/*-template` to the monorepo. But I think many would, as having it in our repo, with multiple maintainers/champions (yeah, we would have champion like setup like in modelina) would make it easier and quicker to reuse solutions from other templates, filters, components and other stuff. Under one repo it would be easier to focus or reusability and good standards (like always using modelina for models generation)

We would still be flexible, independent templates would be possible, hooks and filters too.

Having all under one umbrella would make generator a more vibrant project, more actively developed and visible.

wdyt?

any disadvantages that I do not see?

**Comment by AyushNautiyalDeveloper** [comment]
@derberg I saw we need to implement monorepo for this. I would like to work on this. please let me know how should i go about this.

**Comment by ayushnau** [comment]
@derberg i saw we need to implement the monorepo for this i would like to work on this. Please let know how should i go about this.

**Comment by derberg** [comment]
@ayushnau for now it is proposal and a discussion

**Comment by ayushnau** [comment]
ok @derberg

**Comment by derberg** [comment]
@jonaslagoni @magicmatatjahu wdyt

**Comment by Aryann15** [comment]
Hey! would like to work on this issue, can i get it assigned? @derberg 
Thanks!

**Comment by derberg** [comment]
@ayushnau @Aryann15 are you folks still interested with helping here? did you have experience with introducing monorepo in the past?

first step would be project refactor and migration of first package from https://github.com/asyncapi/generator-filters

if you are interested working on it, please first share your plan before opening a PR

**Comment by ayushnau** [comment]
Hi I would love to work on this issue.

**Comment by derberg** [comment]
added to bounty:

- generator will be now a monorepo, 
- main readme should still be focused on generator and there will just need to be a section where we explain that there are other projects included - short sentence intro and link to dedicated readme
- we need short development guide that will explain the setup of the repo
- filters are the first things we move
- filters repo will be archived 
- we still release `@asyncapi/generator` and we still release `@asyncapi/filters`, just from one repo. Release can have the same number, we do not need separate versioning

any other questions?

**Comment by AyushNautiyalDeveloper** [comment]
<img width="368" alt="image" src="https://github.com/asyncapi/generator/assets/138557911/0bce46f1-7fac-4e4d-9694-8d7e125cd0e3">

**Structure of Monorepo**
1. The Project is divided into mainly two parts apps and packages
   a. Apps contain standalone applications or services, each housed within its own directory. These directories typically include the source code, configuration files, and any assets specific to the application.

  b. Packages encompass shared code, libraries, or modules that are used across multiple apps (in the apps folder) within the monorepo.

2. Other than that we have turbo.json. The turbo.json file serves as a central configuration file for managing the monorepo's setup, dependencies, and tooling. It typically includes metadata about the repository, such as project name, version, and authorship details. 

3. At last we have package.json which serves as a centralized package directory and script files to control all other apps.


**Questions**:
1. Is turborepo final for migrating generator repos to it. 
2. So Generally we put all the functions and miscellaneous stuff like docs in. packages. (package containing all the shared stuff of all the apps.) which can be used so will we take some of shared stuff of the repos and put that into packages.
3. And do we want to put all the ci/cd flows in their own repos or will unite them into one.

**Comment by AyushNautiyalDeveloper** [comment]
@derberg

**Comment by aeworxet** [comment]
### Bounty Issue's service comment

Text labels: `bounty/2024-Q2`, `bounty/advanced`, `bounty/coding`
First assignment to third-party contributors: [2024-03-22 00:00:00 UTC+12:00](https://time.lol/#2024-03-22T00:00:00+12:00)
End Of Life: [2024-08-31 23:59:59 UTC-12:00](https://time.lol/#2024-08-31T23:59:59-12:00)

@asyncapi/bounty_team

**Comment by DhairyaMajmudar** [comment]
@aeworxet I am interested in taking this up : )

**Comment by derberg** [comment]
> Is turborepo final for migrating generator repos to it.

tbh I don't have any preference on what tool we choose. If we do npm workspaces, lerna or turborepo. Afaik you have experience with monorepo, so if you think turborepo is good investment, lets do it.

> So Generally we put all the functions and miscellaneous stuff like common scripts in. packages. (package containing all the shared stuff of all the apps.) which can be used so will we take some of shared stuff of the repos and put that into packages.

we still need to release `@asyncapi/generator` and we will still have to release `@asyncapi/filters`. If there will be some common code then definitely we should have shared commons. 

Filters technically are a feature only for generator templates that run on [Nunjucks](https://github.com/mozilla/nunjucks) engine. So template developer uses it in the way that they add filters package to dependencies and then adds module name to `filters` [in the config](https://www.asyncapi.com/docs/tools/generator/configuration-file). So in theory, your job is a simple as move filters in, make sure they are released and that is it. But would be super nice if you can come up with an idea how to make sure `@asyncapi/filters` are included in `asyncapi/generator` by default, so nobody have to enable them by configuration.

> And do we want to put all the ci/cd flows in their own repos or will unite them into one.

afaik workflows from https://github.com/asyncapi/generator-filters can be dropped as they are all anyway pulled from `.github` repo (so are the same that also in `generator`) repo. You just need to make sure that whatever workflows we have here, for release or bump - will still work

> We want to maintain git history of each app

Not a strong requirement - just lemme know what options we have.
The plan is that basically this https://github.com/asyncapi/generator-filters will be archived and moved to https://github.com/asyncapi-archived-repos so history will still be there. But if there is some cool git magic available so we can preserve history of filters in generator, lets see. I admit I'm not such a git expert


---
Please also work closely with people from https://github.com/asyncapi/parser-js/issues/963 as might be that best would be to have the same solution in both

**Comment by derberg** [comment]
@AyushNautiyalDeveloper so, still interested?

**Comment by AyushNautiyalDeveloper** [comment]
yeah Sure @derberg once you answer those question I will add short development guide. as required.

**Comment by ayushnau** [comment]
yeah Sure @derberg once you answer those question I will add short development guide. as required.

**Comment by derberg** [comment]
I think we have twins here 😄

**Comment by derberg** [comment]
btw, my answers are above

**Comment by derberg** [comment]
@ayushnau ok, please go ahead, the issue is yours

**Comment by ayushnau** [comment]
ok thanks will start working on this then.

**Comment by aeworxet** [comment]
### Bounty Issue's Timeline
|Complexity Level|Assignment date (by GitHub)|Start date (by BP rules)|End date (by BP rules)|Draft PR submission|Final PR submission|Final PR merge|
|----------------|---------------------------|------------------------|----------------------|-------------------|-------------------|--------------|
|Advanced        |2024-03-21                 |[2024-04-01](https://time.lol/#2024-04-01T00:00:00+12:00)              |**[2024-05-24](https://time.lol/#2024-05-24T23:59:59-12:00)**        |[2024-04-19](https://time.lol/#2024-04-19T23:59:59-12:00)         |[2024-05-10](https://time.lol/#2024-05-10T23:59:59-12:00)         |**[2024-05-24](https://time.lol/#2024-05-24T23:59:59-12:00)**    |

##### Please note that the dates given represent deadlines, not specific dates, so if the goal is reached sooner, it's better.
##### Keep in mind the [responsibility](https://github.com/asyncapi/community/pull/897/files#diff-25ecb20a61754c225d6511ca08d7e7c9a14b9ca5a93e89bd42331e51c9ebf26dR201) for violations of the [Timeline](https://github.com/asyncapi/community/pull/897/files#diff-25ecb20a61754c225d6511ca08d7e7c9a14b9ca5a93e89bd42331e51c9ebf26dR127).

**Comment by ayushnau** [comment]
## Comparison of Repository Management Tools

1. So I have Decided to go with Turborepo. Here is a guide for comparing different options we had. Nx was also good.

| Feature                        | TurboRepo                                          | Lerna                                              | Yarn Workspaces                                    | Nx                                                 |
|--------------------------------|----------------------------------------------------|-----------------------------------------------------|-----------------------------------------------------|-----------------------------------------------------|
| Performance                    | High performance with caching and parallelization  | Decent performance, may slow down with large repos  | Decent performance, can suffer with large repos     | High performance with advanced caching and builders|
| Dependency Management          | Integrated dependency resolution                    | Basic dependency management                         | Integrated with Yarn, follows Yarn's resolution     | Integrated dependency graph for efficient builds   |
| Versioning                     | Integrated version management                       | Incremental versioning with fixed or independent    | Integrated with Yarn, follows SemVer               | Semantic versioning with versioning strategies     |
| Language Support               | Primarily JavaScript and TypeScript                | Primarily JavaScript and TypeScript                | JavaScript, TypeScript, and more                   | JavaScript, TypeScript, and more                   |
| Community Support              | Growing community support                           | Large and established community                     | Strong community support                            | Strong community support                            |
| Ease of Setup                  | Straightforward setup and configuration            | Moderate setup with configuration options           | Integrated with Yarn, relatively easy to set up     | Straightforward setup and integration              |
| Integration with CI/CD         | Supports various CI/CD pipelines                   | Compatible with common CI/CD tools                  | Compatible with common CI/CD tools                  | Integrated CI/CD support and workflows              |


**Adding a single repo will look like this**

1. Install and do basic setup of TurboRepo. Note: We will be using npm as a package manager as we have used npm in every package.

2. Add the `@asyncapi/filter` code to the `apps` directory of the turborepo.

3. Add the configuration for the `@asyncapi/filter` to the `Turbo.json` file in the root directory.

4. Add all the required scripts in the `package.json` file for the `@asyncapi/filter`. (To run apps `package.json` scripts from root)

### Question:

1. What should be the name of the repo? It can't be "Generator" as we are going to use it for its own repo.

**Comment by ayushnau** [comment]
https://github.com/ayushnau/generator

**Comment by derberg** [comment]
can you specify what is the goal of https://github.com/ayushnau/generator ? just for testing?
The changes that you introduce should be done in a typical way through a fork

> What should be the name of the repo? It can't be "Generator" as we are going to use it for its own repo.

we keep the name for now, is that a blocker? all additional stuff we will move to `generator`  are considered plugins/extensions - apps is probably a good name too

> So I have Decided to go with Turborepo. Here is a comparision for different options we had

thanks for great comparizon. Turborepo indeed looks interesting. Also notices even nodejs started using it https://github.com/nodejs/nodejs.org

**Comment by ayushnau** [comment]
1. The purpose of that repository is solely for testing and demonstration purposes.Other apps will also be included inside the apps directory (including the generator itself.)
2. No, it is not a blocker,  just asking for clarity.
3. 👍

**Comment by derberg** [comment]
awesome, feel free to DM me whenever you need help or there is something new I should look at - I'm a bit under water with notifications, so DM is working very well for me atm

**Comment by ayushnau** [comment]
https://github.com/asyncapi/generator/pull/1155

**Comment by derberg** [comment]
@ayushnau I noticed you closed the PR, all good? what's the plan?

**Comment by ayushnau** [comment]
@derberg I am getting an issue with running test command. 
<img width="839" alt="image" src="https://github.com/asyncapi/generator/assets/78146753/d808d11a-91f2-4d3e-9b17-17ff1330b70d">
This is coming due to the test trying to access the nims from node_modules. but turborepo have installed it in base node_modules. where all the internal and shared dependencies packages are installed. 

Solution:
I think what we can do is just change the directory 

```js
 '^nimma/legacy$': '<rootDir>/node_modules/nimma/dist/legacy/cjs/index.js',
    '^nimma/(.*)': '<rootDir>/node_modules/nimma/dist/cjs/$1', 
```
to correct directory.

**Comment by derberg** [comment]
yup, go ahead 😀

**Comment by aeworxet** [comment]
@ayushnau, please provide an update to the [PR](https://github.com/asyncapi/generator/pull/1186).

**Comment by ayushnau** [comment]
@aeworxet adding

**Comment by derberg** [comment]
fyi, the implementation of the issue is delayed a bit as I was out for a couple of weeks for conference and holidays - so could not review, and there was noone else who could

**Comment by derberg** [comment]
not yet completed, we need new release pipeline and move into project one more app

**Comment by ayushnau** [comment]
@derberg can we mark this issue completed

**Comment by derberg** [comment]
@aeworxet bounty scope completed. I'm not closing the issue as there is still a need to move hooks into generator - but that was not part of bounty scope

**Comment by aeworxet** [comment]
After a delay that was beyond the control of the Bounty Program Participant
https://github.com/asyncapi/generator/issues/1044#issuecomment-2178009805

### Bounty Issue Completed 🎉

@ayushnau, please go to the [AsyncAPI page on Open Collective](https://opencollective.com/asyncapi) and submit an invoice for `USD 400.00` with the expense title `Bounty generator#1044`, tag `bounty`, and full URL of this Bounty Issue in the description.

**Comment by derberg** [comment]
this should not yet be closed, `generator-react-sdk` is not yet moved here, we need to chat with the maintainers of this project, if they are ok that we move it here to generator - probably a GH issue enquiry has to start

**Comment by mukeshgupta7** [comment]
Is bounty still open @derberg

**Comment by aeworxet** [comment]
@mukeshgupta7 

This Bounty Issue has been completed ten months ago:
https://github.com/asyncapi/generator/issues/1044#issuecomment-2263050162
https://github.com/asyncapi/generator/issues/1044#issuecomment-2551521415

However, looking [at the previous PRs](https://github.com/orgs/asyncapi/projects/36/views/4?filterQuery=&sliceBy%5BcolumnId%5D=78904519&sliceBy%5Bvalue%5D=generator%231044), you are welcome to evaluate whether you would be able to complete this issue if assigned to it, and to choose an issue of a similar complexity level within the https://github.com/asyncapi/generator repository if so.

**Comment by aeworxet** [comment]
### Bounty Issue's service comment

Text labels: `bounty/2025-Q3`, `bounty/medium`, `bounty/coding`
First assignment to regular contributors: [2025-06-20 00:00:00 UTC+12:00](https://time.lol/#2025-06-20T00:00:00+12:00)
End Of Life after: [2025-07-31 23:59:59 UTC-12:00](https://time.lol/#2025-07-31T23:59:59-12:00)

@asyncapi/bounty_team

##### The Bounty Program is not a Mentorship Program. The accepted level of Bounty Program Participants is **Middle**/**Senior**.
##### Regular contributors should explain in meaningful words how they are going to approach the resolution process when expressing a desire to work on this Bounty Issue.

**Comment by derberg** [comment]
I'll be working on it

**Comment by aeworxet** [comment]
Following the [expression of desire to work on this Bounty Issue](https://github.com/asyncapi/generator/issues/1044#issuecomment-2983362900), @derberg (`githubID`: [6995927](https://api.github.com/user/6995927)) is an AsyncAPI Maintainer specified in the file https://github.com/asyncapi/community/blob/master/MAINTAINERS.yaml, so they fall under the first category in the [prioritization list](https://github.com/asyncapi/community/blob/master/docs/010-contribution-guidelines/BOUNTY_PROGRAM.md?plain=1#L33).

**Comment by aeworxet** [comment]
### Bounty Issue's Timeline
|Complexity Level|Assignment Date (by GitHub)|Start Date (by BP Rules)|End Date (by BP Rules)|Draft PR Submission|Final PR Merge Start|Final PR Merge End|
|----------------|---------------------------|------------------------|----------------------|-------------------|--------------------|------------------|
|Medium          |2025-06-18              |[2025-07-07](https://time.lol/#2025-07-07T00:00:00+12:00)              |**[2025-08-17](https://time.lol/#2025-08-17T23:59:59-12:00)**        |[2025-07-20](https://time.lol/#2025-07-20T23:59:59-12:00)         |[2025-08-03](https://time.lol/#2025-08-03T23:59:59-12:00)         |**[2025-08-17](https://time.lol/#2025-08-17T23:59:59-12:00)**        |

##### Please note that the dates given represent deadlines, not specific dates; so if the goal is reached sooner, it's better.
##### Keep in mind the [responsibility](https://github.com/asyncapi/community/blob/master/docs/010-contribution-guidelines/BOUNTY_PROGRAM.md?plain=1#L223) for violations of the [Timeline](https://github.com/asyncapi/community/blob/master/docs/010-contribution-guidelines/BOUNTY_PROGRAM.md?plain=1#L96).
#
Assignee: @derberg (`githubID`: [6995927](https://api.github.com/user/6995927))

**Comment by derberg** [comment]
react-sdk is now part of the monorepo

<img width="1420" height="976" alt="Image" src="https://github.com/user-attachments/assets/937f74a2-e3ec-4975-a98c-81c625c2d3d7" />

next step for me is to archive https://github.com/asyncapi/generator-react-sdk/ with proper note where the code is now

**Comment by derberg** [comment]
https://github.com/asyncapi-archived-repos/generator-react-sdk archived and moved

@aeworxet this task is completed now

**Comment by aeworxet** [comment]
### Bounty Issue Is Completed 🎉

@derberg (`githubID`: [6995927](https://api.github.com/user/6995927)), please go to the dedicated AsyncAPI Bounty Program `2025-Q3` [page on Open Collective](https://opencollective.com/asyncapi/projects/asyncapi-bounty-program) and submit an invoice for `USD 200.00` (button '`ACTIONS`', dropdown option '`Submit expense`') with the expense title `Bounty generator#1044`, tag `bounty`, and full URL of this Bounty Issue in the description.

After submitting the invoice, please post the link to it in this Bounty Issue as a separate comment to verify the invoice's authorship.

**Comment by derberg** [comment]
https://opencollective.com/asyncapi/projects/asyncapi-bounty-program/expenses/262764

**Comment by aeworxet** [comment]
`derberg`
`https://opencollective.com/asyncapi/projects/asyncapi-bounty-program/expenses/262764`

✅
The invoice https://opencollective.com/asyncapi/projects/asyncapi-bounty-program/expenses/262764 was submitted by @derberg (`githubID`: [6995927](https://api.github.com/user/6995927)), who was the AsyncAPI Bounty Program `2025-Q3` Participant and completed the Bounty Issue `generator#1044`.
